src/core/mesh/qgstriangularmesh.cpp

```diff
@@ -0,0 +1,110 @@
+/***************************************************************************
+                         qgstriangularmesh.cpp
+                         ---------------------
+    begin                : April 2018
+    copyright            : (C) 2018 by Peter Petrik
+    email                : zilolv at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgstriangularmesh.h"
+#include "qgsmeshdataprovider.h"
+#include "qgsrendercontext.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgscoordinatetransform.h"
+#include "qgsmeshlayer.h"
+
+
+QgsTriangularMesh::QgsTriangularMesh( )
+{
+}
+
+QgsTriangularMesh::~QgsTriangularMesh()
+{
+}
+
+void QgsTriangularMesh::update( QgsMesh *nativeMesh, QgsRenderContext *context )
+{
+  Q_ASSERT( nativeMesh );
+  Q_ASSERT( context );
+
+  mTriangularMesh.vertices.clear();
+  mTriangularMesh.faces.clear();
+  mTrianglesToNativeFaces.clear();
+
+  // TRANSFORM VERTICES
+  QgsCoordinateTransform transform = context->coordinateTransform();
+  mTriangularMesh.vertices.resize( nativeMesh->vertices.size() );
+  for ( int i = 0; i < nativeMesh->vertices.size(); ++i )
+  {
+    QgsMeshVertex vertex = nativeMesh->vertices[i];
+    if ( transform.isValid() )
+    {
+      QgsPointXY mapPoint = transform.transform( QgsPointXY( vertex.x(), vertex.y() ) );
+      QgsMeshVertex mapVertex( mapPoint );
+      mapVertex.setZ( vertex.z() );
+      mapVertex.setM( vertex.m() );
+      mTriangularMesh.vertices[i] = mapVertex;
+    }
+    else
+    {
+      mTriangularMesh.vertices[i] = vertex;
+    }
+  }
+
+  // CREATE TRIANGULAR MESH
+  for ( int i = 0; i < nativeMesh->faces.size(); ++i )
+  {
+    QgsMeshFace face = nativeMesh->faces[i] ;
```

Comment:
this may detach - use faces.at( i ) instead (and use Qt creator 4.6 with all levels of clazy warnings enabled to catch these)